File: mesop/components/form_field/form_field.py
from typing import Literal

from pydantic import validate_arguments

import mesop.components.form_field.form_field_pb2 as form_field_pb
from mesop.component_helpers import (
  insert_composite_component,
)


@validate_arguments
def form_field(
  *,
  key: str | None = None,
  hide_required_marker: bool = False,
  color: Literal["primary", "accent", "warn"] = "primary",
  float_label: Literal["always", "auto"] = "always",
  appearance: Literal["fill", "outline"] = "fill",
  subscript_sizing: Literal["fixed", "dynamic"] = "fixed",
  hint_label: str = "",
):
  """
  TODO_doc_string
  """
  return insert_composite_component(
    key=key,
    type_name="form_field",
    proto=form_field_pb.FormFieldType(
      hide_required_marker=hide_required_marker,
      color=color,
      float_label=float_label,
      appearance=appearance,
      subscript_sizing=subscript_sizing,
      hint_label=hint_label,
    ),
  )

File: mesop/components/form_field/form_field.ts
import {
  MatFormField,
  MatFormFieldControl,
  MatFormFieldModule,
} from '@angular/material/form-field';
import {Component, ContentChild, Input, ViewChild} from '@angular/core';
import {
  UserEvent,
  Key,
  Type,
} from 'mesop/mesop/protos/ui_jspb_proto_pb/mesop/protos/ui_pb';
import {FormFieldType} from 'mesop/mesop/components/form_field/form_field_jspb_proto_pb/mesop/components/form_field/form_field_pb';
import {Channel} from '../../web/src/services/channel';
import {MatInputModule} from '@angular/material/input';

@Component({
  templateUrl: 'form_field.ng.html',
  standalone: true,
  imports: [MatFormFieldModule, MatInputModule],
})
export class FormFieldComponent {
  @Input({required: true}) type!: Type;
  @Input() key!: Key;
  private _config!: FormFieldType;

  @ContentChild(MatFormFieldControl) _control!: MatFormFieldControl<any>;
  @ViewChild(MatFormField) _matFormField!: MatFormField;

  ngOnInit() {
    this._matFormField._control = this._control;
  }

  constructor(private readonly channel: Channel) {}

  ngOnChanges() {
    this._config = FormFieldType.deserializeBinary(
      this.type.getValue() as unknown as Uint8Array,
    );
  }

  config(): FormFieldType {
    return this._config;
  }

  getColor(): 'primary' | 'accent' | 'warn' {
    return this.config().getColor() as 'primary' | 'accent' | 'warn';
  }

  getFloatLabel(): 'always' | 'auto' {
    return this.config().getFloatLabel() as 'always' | 'auto';
  }

  getAppearance(): 'fill' | 'outline' {
    return this.config().getAppearance() as 'fill' | 'outline';
  }

  getSubscriptSizing(): 'fixed' | 'dynamic' {
    return this.config().getSubscriptSizing() as 'fixed' | 'dynamic';
  }
}
